src/shaders/pbr.frag

```diff
@@ -0,0 +1,246 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+// This is an implementation of
+// Karis, Brian. “Real Shading in Unreal Engine 4 by.” (2013).
+
+// -------------- input ---------------------
+// position, normal, tangent in camera space
+in highp vec3 position;
+in highp vec3 normal;
+#if defined(TEXTURED)
+in mediump vec2 texCoord;
+#endif
+#if defined(NORMAL_TEXTURE) && defined(PRECOMPUTED_TANGENT)
+in highp vec3 tangent;
+in highp vec3 biTangent;
+#endif
+
+// -------------- output -------------------
+layout(location = 0) out vec4 fragmentColor;
+#if defined(OBJECT_ID)
+layout(location = 1) out highp uint fragmentObjectId;
+#endif
+
+// -------------- lights -------------------
+struct LightData {
+  // WARNING:
+  // In this shader, the light intensity is considered in the lightColor!!
+  vec3 lightColor;
+  float range;
+};
+uniform LightData Light[LIGHT_COUNT];
+
+// if .w == 0, it means it is a dirctional light, .xyz is the direction;
+// if .w == 1, it means it is a point light, .xyz is the light position;
+// it is NOT put in the Light Structure, simply because we may modify the code
+// so it is computed in the vertex shader.
+uniform vec4 LightDirections[LIGHT_COUNT];
+
+// -------------- material, textures ------------------
+struct MaterialData {
+  vec4 baseColor;   // diffuse color, if albedoTexture exists,
+                    // use the albedoTexture
+  float roughness;  // roughness of a surface, if roughness texture exists,
+                    // use the roughnessTexture
+  float metallic;   // metalness of a surface, if metallic texture exists,
+                    // use the metallicTexture
+};
+uniform MaterialData Material;
+
+#if defined(ALBEDO_TEXTURE)
+uniform sampler2D albedoTexture;
+#endif
+#if defined(ROUGHNESS_TEXTURE)
+uniform sampler2D roughnessTexture;
+#endif
+#if defined(METALLIC_TEXTURE)
+uniform sampler2D metallicTexture;
+#endif
+#if defined(NORMAL_TEXTURE)
+uniform sampler2D normalTexture;
+#endif
+
+// -------------- uniforms ----------------
+#if defined(OBJECT_ID)
+uniform highp uint objectId;
+#endif
+
+// -------------- shader ------------------
+#if defined(NORMAL_TEXTURE)
+vec3 getNormalFromNormalMap() {
+  vec3 tangentNormal = texture(normalTexture, texCoord).xyz * 2.0 - 1.0;
+
+#if defined(PRECOMPUTED_TANGENT)
+  mat3 TBN = mat3(tangent, biTangent, normal);
+#else
+// TODO: some magic here!
```

Comment:
`#error`? because otherwise, until this is implemented, you'll get all-black models and you wouldn't know why